File: 2025-1/sesion-de-clase-20/Restaurante.cpp
/* 
 * File:   Restaurante.cpp
 * Author: erichuiza
 * 
 * Created on May 2, 2025, 5:21 PM
 */

#include "Restaurante.h"
#include "Cliente.h"
#include "PedidoRegular.h"
#include "PedidoPrioritario.h"
#include "PedidoInfo.h"
#include "Comparadores.h"


void Restaurante::cargarMeseros(const string nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    
    Mesero mesero;
    while (archivo >> mesero) {
        this->meseros.push_back(mesero);
    }
    
    sort(this->meseros.begin(), 
            this->meseros.end(), 
            OrdenarMeseroPorExperiencia());
}

void Restaurante::cargarClientes(const string nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    
    Cliente cliente;
    while (archivo >> cliente) {
        this->clientes.push_back(cliente);
    }
}

void Restaurante::cargarPedidos(const string nombreArchivo) {
    ifstream archivo(nombreArchivo, ios::in);
    
    int prioridad;
    while (!archivo.eof() && archivo >> prioridad) {
        archivo.ignore();
        
        Pedido *pedido;
        switch (prioridad) {
            case 0:
                pedido = new PedidoRegular();
                break;
            case 1:
                pedido = new PedidoPrioritario();
                break;
        }
        
        if (archivo >> *pedido) {
            PedidoInfo pedidoInfo(pedido);
            this->colaPedidos.push(pedidoInfo);
        }
    }
}

void Restaurante::reportePedidos(const string nombreArchivo) {
    ofstream archivo(nombreArchivo, ios::out);
    
    archivo 
        << left << setw(10) << "PRIORITARIO"
        << left << setw(15) << "CODIGO"
        << setw(15) << "DNI"
        << setw(50) << "PLATO"
        << endl;
    
    // Copio la cola porque la cola prioritaria de la STL no tiene
    // iteradores directamente, así que copiamos para mantener la cola de pedidos 
    // intacta.
    auto copia = this->colaPedidos;
    while (!copia.empty()) {
        const PedidoInfo& pedido = copia.top();
        archivo << *pedido.getPedido();
        
        copia.pop();
    }
}

void Restaurante::reporteMeseros(const string nombreArchivo) {
    ofstream os(nombreArchivo, ios::out);
    
    os 
        << left << setw(15) << "DNI"
        << setw(50) << "NOMBRE"
        << setw(15) << "SUELDO"
        << setw(15) << "EXPERIENCIA"
        << endl;

    for (auto it = 
            this->meseros.begin();it != 
            this->meseros.end(); it++) {
        os << *it;
    }
}

void Restaurante::reporteClientes(const string nombreArchivo) {
    ofstream os(nombreArchivo, ios::out);
    
    os 
        << left << setw(15) << "DNI"
        << setw(50) << "NOMBRE"
        << setw(20) << "TELEFONO"
        << setw(20) << "CORREO"
        << endl;

    for (vector<Cliente>::iterator it = this->clientes.begin();it != this->clientes.end(); it++) {
        os << *it;
    }
}

File: 2025-2/sesion-de-clase-26-1/src/Restaurante.java
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 *
 * @author erichuiza
 */
public class Restaurante {
    private final List<Mesero> meseros;
    private final List<Cliente> clientes;
    private final List<Pedido> pedidos;

    public Restaurante() {
        this.meseros = new ArrayList<>();
        this.clientes = new ArrayList<>();
        this.pedidos = new ArrayList<>();
    }
    
    public void cargarMeseros(final String nombreArchivo) throws Exception {
        File archivo = new File(nombreArchivo);
        Scanner scanner = new Scanner(archivo);
        scanner.useDelimiter(",|\\n");
        
        while (scanner.hasNext()) {
            Mesero mesero = new Mesero();
            mesero.cargarDatos(scanner);
            this.meseros.add(mesero);
        }
    }
    
    public void imprimirMeseros() {
        for (Mesero mesero : this.meseros) {
            mesero.imprimir();
            System.out.println();
        }
    }
    
    public void cargarClientes(final String nombreArchivo) throws Exception {
        File archivo = new File(nombreArchivo);
        Scanner scanner = new Scanner(archivo);
        scanner.useDelimiter(",|\\n");
        
        while (scanner.hasNext()) {
            Cliente cliente = new Cliente();
            cliente.cargarDatos(scanner);
            this.clientes.add(cliente);
        }
    }
    
    public void imprimirClientes() {
        for (Cliente cliente : this.clientes) {
            cliente.imprimir();
            System.out.println();
        }
    }
    
    public void cargarPedidos(final String nombreArchivo) throws Exception {
        File archivo = new File(nombreArchivo);
        Scanner scanner = new Scanner(archivo);
        scanner.useDelimiter(",|\\n");
        
        while (scanner.hasNext()) {
            Pedido pedido;
            int tipo = Integer.parseInt(scanner.next());
            
            switch (tipo) {
                case 0: 
                    pedido = new PedidoRegular();
                    break;
                case 1: 
                    pedido = new PedidoPrioritario();
                    break;
                default:
                    pedido = new PedidoRegular();
                    break;
            }
            
            pedido.cargarDatos(scanner);
            this.pedidos.add(pedido);
        }
    }
    
    public void imprimirPedidos() {
        for (Pedido pedido : this.pedidos) {
            pedido.imprimir();
            System.out.println();
        }
    }
}
